src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -906,5 +914,141 @@ private static void ReadToEndOfCurrentObject(JsonReader reader)
 
             throw new JsonSerializationException($"Could not read till end of object, end of stream. Got '{reader.TokenType}'.");
         }
+
+        private static string Intern(string str)
+        {
+            if (str is null)
+            {
+                return null;
+            }
+
+            var interned = string.IsInterned(str);
+            if (interned != null)
+            {
+                return interned;
+            }
+
+            return weakReferenceTable.Intern(str);
+        }
+
+        /// <summary>
+        /// The purpose of this class is to avoid permanently storing duplicate strings which were read in from JSON.
+        /// </summary>
+        private class WeakReferenceTable
+        {
+            public static readonly WeakReferenceTable Instance = new WeakReferenceTable();
+
+            private WeakReferenceTable()
+            {
+            }
+
+            private readonly WeakValueDictionary<string, string> _table = new WeakValueDictionary<string, string>();
+
+            public string Intern(string str)
+            {
+                if (_table.TryGetValue(str, out var value))
+                {
+                    return value;
+                }
+                else
+                {
+                    _table.Add(str, str);
```

Comment:
This interaction currently isn't thread proof. If we have two different deserializations happening simultaneously you can potentially have two "adds" occurring simultaneously meaning the string you're trying to insert here may not be the one "cached" resulting in an extra allocation.

In general it's interesting that the `_table` itself is "kind of " thread safe but this interaction isn't. I'd almost rather have the underlying `WeakValuedDictionary` not thread safe and make the top-level piece here thread safe.

Reply:
Good catch, thinking about thread-safety was one of my last steps there near the end of the day and clearly it needed more bake-time.